watchmal/engine/engine_classifier.py

```diff
@@ -206,130 +230,294 @@ def train(self, train_config):
+                            global_val_metrics = local_val_metrics
+
+                        if self.rank == 0:
+                            # Save if this is the best model so far
+                            global_val_loss = np.mean(global_val_metrics["loss"])
+                            global_val_accuracy = np.mean(global_val_metrics["accuracy"])
+
+                            val_metrics["loss"] = global_val_loss
+                            val_metrics["accuracy"] = global_val_accuracy
+
+                            if val_metrics["loss"] < best_val_loss:
+                                print('best validation loss so far!: {}'.format(best_val_loss))
+                                self.save_state(best=True)
+                                val_metrics["saved_best"] = 1
+
+                                best_val_loss = val_metrics["loss"]
+
+                            # Save the latest model if checkpointing
+                            if checkpointing:
+                                self.save_state(best=False)
+
+                            self.val_log.record(val_metrics)
+                            self.val_log.write()
+                            self.val_log.flush()
+
+                    # Train on batch
+                    self.data = data
+                    self.labels = self.data.y
+
+                    # Call forward: make a prediction & measure the average error using data = self.data
+                    res = self.forward(True)
+
+                    #Call backward: backpropagate error and update weights using loss = self.loss
+                    self.backward()
+
+                    # update the epoch and iteration
+                    epoch          += 1./len(self.data_loaders["train"])
+                    self.iteration += 1
+
+                    # get relevant attributes of result for logging
+                    train_metrics = {"iteration": self.iteration, "epoch": epoch, "loss": res["loss"], "accuracy": res["accuracy"]}
+
+                    # record the metrics for the mini-batch in the log
+                    self.train_log.record(train_metrics)
+                    self.train_log.write()
+                    self.train_log.flush()
+
+                    # print the metrics at given intervals
+                    if self.rank == 0 and self.iteration % report_interval == 0:
+                        previous_iteration_time = iteration_time
+                        iteration_time = time()
+                        print("... Iteration %d ... Epoch %1.2f ... Training Loss %1.3f ... Training Accuracy %1.3f ... Time Elapsed %1.3f ... Iteration Time %1.3f" %
+                              (self.iteration, epoch, res["loss"], res["accuracy"], iteration_time - start_time, iteration_time - previous_iteration_time))
+
+                    if epoch >= epochs:
+                        break
+
+            self.train_log.close()
+            if self.rank == 0:
+                self.val_log.close()
+
+    
+    
```

Comment:
Looks like a big block of copied and pasted code here with just a few differences. This should be avoided as it makes if much more difficult to maintain. Could you merge these together, then e.g. just use switches at the places where they differ? Similarly for other blocks below, I think.